---
title: "HLA imputation report"
date: "`r Sys.Date()`"
output: 
  bookdown::html_document2:
    citation_package: natbib
    fig_width: 4 
    fig_height: 4
    number_sections: true
    fig_caption: true
    toc: true
editor_options:
  chunk_output_type: console
bibliography: bib/bibtex.bib
link-citations: yes
params:
  checked_name : "test"
  rootname : "test"
  pop : "African American"
  model: "multiethnic_IKMB.RData"
  shapeit: "no shapeit"
  modules: "no modules"
  assembly: "assembly"
  basenamerunname: "test"
header-includes:
   - \usepackage[english]{babel}
   - \usepackage{hyperref}
   - \usepackage{caption}
---
\fontsize{10}{15}
\selectfont


```{r style, echo = F, results="asis", cache=FALSE, message = F, warning=FALSE}

# Set knit global options
library("knitr")
library("rmarkdown")
library("ggplot2")
library("pdftools")
library("magick")


options(digits = 2, width = 80)
golden_ratio <- (1 + sqrt(5)) / 2
opts_chunk$set(echo = TRUE,
               tidy = FALSE,
               include = TRUE,
               cache = FALSE)
opts_knit$set(eval.after = "fig.cap")


model = gsub(".*models/", "", params$model)
options(stringsAsFactors=F)

```


# Details of the imputation

Reference model used
```{r refcheck info, echo =FALSE,  results="markup",cache=FALSE, message = F}
# ALIGNMENT TO MODEL
model

```


## Alignment of input data to reference
Figures 1.1 and 1.2 show the alignment of the reference to the input data based on allele frequency and the overlap of positions between the reference and the input data.
<br>
```{r figure1, echo = F, out.width="80%",results= "markup",  cache=FALSE, message = F, fig.cap=c("Alignment cohort to reference BEFORE and AFTER SNPs were excluded or flipped."), fig.pos="!ht"}

file = image_read_pdf(paste0("refchecked_A_",params$basenamerunname,".pdf"))
par(mar=c(0,0,0,0))
plot(file)
```
<br>
```{r figure2, echo = F, out.width="80%", results="markup", cache=FALSE, message = F, fig.cap=c("Distribution of % of panel SNPs used per classifier."), fig.pos="!ht"}
file= image_read_pdf(paste0("refchecked_B_",params$basenamerunname,".pdf") )
par(mar=c(0,0,0,0))
plot(file)
```

## Imputation accuracy
Figure 1.3 shows the distribution of posterior probability of predicted genotypes (i.e for both predicted HLA alleles per locus). It shows the general imputation quality per gene. Figure 1.4 shows the marginal posterior probability per single allele and locus. 
<br>
```{r figure3, out.width="80%", echo = F, results="markup", cache=FALSE, message = F, fig.cap="The distribution of posterior probability of predicted genotypes given by HIBAG [@Zheng2014]", fig.pos="!ht"}
file= image_read_pdf(paste0("postprob_",params$basenamerunname,".pdf"))
par(mar=c(0,0,0,0))
plot(file)
```
<br>

```{r figure4, out.width="80%", echo = F,  results="markup", cache=FALSE, message = F, fig.cap=" The distribution of marginal posterior probabilities for each allele per locus [@Degenhardt2021]", fig.pos="!ht"}
file= image_read_pdf(paste0("unsure_",params$basenamerunname,".pdf"))
par(mar=c(0,0,0,0))
plot(file)
```


## phaseHLACombine: Phasing certainty
Figure 1.5 gives and overview of the quality of the phasing with certainty ranging from 0 to 1. 
<br>
```{r figure5, echo = F, results="markup",  out.width="80%", cache=FALSE, message = F, fig.cap="Phasing certainty across the loci. Calculated across 10 classifiers.  [@Degenhardt2021]", fig.pos="!ht"}
file= image_read_pdf(paste0("phased_",params$basenamerunname,".pdf"))
par(mar=c(0,0,0,0))
plot(file)
```

```{r get PHASEOVERLAP, echo = F,  results="hide", cache=FALSE, message = F}
res = data.frame(read.table(paste0("imputation_",params$checked_name,".META.PHASING.txt"), h=T,sep="\t"))
N = table(paste(res$locus, res$true.geno)) # true N of allele combination
res = res[res$phase_prob < 0.6,]
res = table(paste(res$locus, res$true.geno)) 
out = cbind(round(res/N[names(res)],4)*100, N[names(res)])

tmp = do.call(rbind, strsplit(names(res)," "))
alleles = cbind(AlleleA=paste0(tmp[,1],"*",tmp[,2]), 
		AlleleB=paste0(tmp[,1], "*", tmp[,3]))
alleles = data.frame(alleles)
frq = read.table(paste0("imputation_",params$checked_name,".info"), h=T)
freq = data.frame(freqAlleleA=frq$AF_ALL[match(alleles$AlleleA,gsub("imputed_HLA_", "",frq$id))],
                  freqAlleleB=frq$AF_ALL[match(alleles$AlleleB,gsub("imputed_HLA_", "",frq$id))])
colnames(out)=c("%","N")
out = cbind(alleles, freq, out)
out = out[out[,"N"]>5,]

```


## Lookup of measures gives an idea about the accuracies of the imputed allele
This table shows potential problematic HLA alleles based on the SNP sequences for alleles stored in the reference dataset. Here positions that overlap between the input data and the reference were extracted and the overlap (i.e. Hemming distance = 0) between SNP haplotypes of different alleles was calculated across all classifiers (typically 100) present in the model. This table should give an idea of which alleles may be mistaken for other alleles by the model.  NClassifiers denotes in how many classifiers both alleles of an allele combination were present. One classifier may contain multiple SNP haplotypes per allele. The %- overlap states how many haplotypes stored for two alleles overlapped within one classifier. Quantiles of these overlaps are also given.

```{r, echo = F,  results="hide", cache=FALSE, message = F}
res = data.frame(read.table(paste0("imputation_overlap_alleles_", params$checked_name, ".txt"), h=T,sep="\t"))
colnames(res) = c( "AlleleA","AlleleB", "freqAlleleA", "freqAlleleB", "0%", "25%", "50%", "75%", "100%", "NClassifiers")
rownames(res)=NULL
```

```{r, echo = F,  results="asis", cache=FALSE, message = F}
kable(res[res[,5] > 5,],digits = 3, caption = "Alleles for which the SNP haplotypes overlap (given the positions in the .bim file) is at minimum > 5%. Quantiles of % overlap are shown across NClassifieres. [@Degenhardt2019]")# format.args = list(scientific = TRUE),
```


```{r, echo = F,  results="hide", cache=FALSE, message = F}


res = data.frame(read.table(paste0("imputation_marginal_prob_",params$checked_name,".txt"), h=T))
res=res[res$prob < 0.8 & res$freq >= 0.01,]
rownames(res)=NULL
```

This table shows potential problematic HLA alleles based on marginal posterior probabilities of the imputation. 
```{r, echo = F,  results="asis", cache=FALSE, message = F}
kable(res,digits = 3, caption = "Alleles with a marginal posterior probability < 0.8 and an allele frequency over 1%.")
```

This table shows which allele combinations were generally difficult to phase.
```{r, echo = F,  results="asis", cache=FALSE, message = F}
rownames(out) = NULL
kable(out, caption = "% Allele combinations with phasing probability < 0.6 and total number of individuals carrying this allele combination (N)).")# format.args = list(scientific = TRUE),
```

# Software

## R-packages
```{r sessioninfo, echo = F,  results="hide", cache=FALSE, message = F, warning=F}

library(parallel)
library(ggplot2)
library(data.table)
library(reshape)
library(HIBAG)
library(ggplot2)
library(grid)
library(gridExtra)
library(reshape2)

```

```{r sessioninfo out, echo = TRUE,  results="markup", cache=FALSE, message = F}
sessionInfo()

```
## CLUSTER MODULES

```{r, echo =F, results="markup"}
unlist(strsplit(params$modules, ":"))
``` 

## SHAPEIT
```{r, echo =F, results="markup"}
Sys.setenv(SHAPEIT=params$shapeit)                                                                                       
```      
```{bash, echo =F, results="markup"}
$SHAPEIT --version
```

## PLINK
```{bash, echo =F, results="markup"}      
plink --version                                                                                                                             
```  

# References
